File: frontend_mobile/src/services/transacciones.ts
/**
 * Hotel Munich - Transaccion Service
 * ====================================
 * Handles payment transaction creation, voiding, and saldo queries.
 */

import { apiGet, apiPost } from './api';

export type PaymentMethod = 'EFECTIVO' | 'TRANSFERENCIA' | 'POS';

export interface Transaccion {
    id: number;
    reserva_id: string | null;
    caja_sesion_id: number | null;
    amount: number;
    payment_method: PaymentMethod;
    reference_number: string | null;
    description: string | null;
    created_at: string;
    created_by: string | null;
    voided: boolean;
    void_reason: string | null;
    voided_at: string | null;
    voided_by: string | null;
}

export interface RegistrarPagoRequest {
    reserva_id: string;
    amount: number;
    payment_method: PaymentMethod;
    reference_number?: string;
    description?: string;
}

export interface AnularRequest {
    reason: string;
}

export interface SaldoReserva {
    reserva_id: string;
    total: number;
    paid: number;
    pending: number;
    transacciones: Transaccion[];
}

/**
 * Register a payment against a reservation.
 * EFECTIVO requires an open caja session for the logged-in user.
 */
export async function registrarPago(data: RegistrarPagoRequest): Promise<Transaccion> {
    return apiPost<Transaccion>('/transacciones/', data);
}

/**
 * Void (anular) a transaction. Reason must be at least 3 characters.
 * Both admin and recepcion can void, action is audited.
 */
export async function anularTransaccion(
    transaccionId: number,
    reason: string
): Promise<Transaccion> {
    return apiPost<Transaccion>(`/transacciones/${transaccionId}/anular`, { reason });
}

/**
 * Get saldo (total/paid/pending) and all active transactions for a reservation.
 */
export async function getSaldoReserva(reservaId: string): Promise<SaldoReserva> {
    return apiGet<SaldoReserva>(`/reservations/${reservaId}/saldo`);
}

/**
 * Get saldo via the transacciones endpoint (same result).
 */
export async function getSaldoFromTransacciones(reservaId: string): Promise<SaldoReserva> {
    return apiGet<SaldoReserva>(`/transacciones/reserva/${reservaId}`);
}

/**
 * List all transactions with optional filters.
 */
export async function listTransacciones(params?: {
    dateFrom?: string;
    dateTo?: string;
    paymentMethod?: PaymentMethod;
    includeVoided?: boolean;
}): Promise<Transaccion[]> {
    const query = new URLSearchParams();
    if (params?.dateFrom) query.append('date_from', params.dateFrom);
    if (params?.dateTo) query.append('date_to', params.dateTo);
    if (params?.paymentMethod) query.append('payment_method', params.paymentMethod);
    if (params?.includeVoided) query.append('include_voided', 'true');
    const q = query.toString() ? `?${query.toString()}` : '';
    return apiGet<Transaccion[]>(`/transacciones/${q}`);
}

/**
 * Get display label for a payment method.
 */
export function paymentMethodLabel(method: PaymentMethod): string {
    const map: Record<PaymentMethod, string> = {
        EFECTIVO: 'Efectivo',
        TRANSFERENCIA: 'Transferencia',
        POS: 'POS (Tarjeta)',
    };
    return map[method] || method;
}

/**
 * Get emoji for a payment method.
 */
export function paymentMethodEmoji(method: PaymentMethod): string {
    const map: Record<PaymentMethod, string> = {
        EFECTIVO: '💵',
        TRANSFERENCIA: '🏦',
        POS: '💳',
    };
    return map[method] || '💰';
}

File: backend/api/v1/endpoints/transacciones.py
"""
Hotel Munich — Transacciones Endpoints
========================================
Register and void payment transactions against reservations.
"""

from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import Optional, List
from datetime import datetime

from api.deps import get_db, get_current_user
from services import TransaccionService, TransaccionError
from schemas import (
    TransaccionCreate,
    AnularTransaccionRequest,
)
from logging_config import get_logger

logger = get_logger(__name__)

router = APIRouter()


def _trans_to_dto(t) -> dict:
    return {
        "id": t.id,
        "reserva_id": t.reserva_id,
        "caja_sesion_id": t.caja_sesion_id,
        "amount": t.amount,
        "payment_method": t.payment_method,
        "reference_number": t.reference_number,
        "description": t.description,
        "created_at": t.created_at,
        "created_by": t.created_by,
        "voided": t.voided,
        "void_reason": t.void_reason,
        "voided_at": t.voided_at,
        "voided_by": t.voided_by,
    }


@router.post("/", summary="Registrar pago")
def registrar_pago(
    data: TransaccionCreate,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user),
):
    """
    Register a payment against a reservation.
    EFECTIVO requires an open caja session for the current user.
    """
    try:
        trans = TransaccionService.registrar_pago(
            db,
            reserva_id=data.reserva_id,
            amount=data.amount,
            payment_method=data.payment_method,
            reference_number=data.reference_number,
            description=data.description,
            created_by=current_user.username,
            user_id=current_user.id,
        )
        return _trans_to_dto(trans)
    except TransaccionError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.post("/{transaccion_id}/anular", summary="Anular una transaccion")
def anular(
    transaccion_id: int,
    data: AnularTransaccionRequest,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user),
):
    """
    Void a transaction. Both admin and recepcion can void, but reason is mandatory.
    The transaction is not deleted — marked voided with reason/user/timestamp.
    """
    user_role = (current_user.role or "").lower().strip()
    if user_role not in ("admin", "supervisor", "gerencia", "recepcion", "recepcionista"):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="No tiene permisos para anular transacciones",
        )

    try:
        trans = TransaccionService.anular_transaccion(
            db,
            transaccion_id=transaccion_id,
            reason=data.reason,
            user=current_user.username,
        )
        return _trans_to_dto(trans)
    except TransaccionError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.get("/", summary="Listar transacciones")
def list_transactions(
    date_from: Optional[datetime] = Query(default=None),
    date_to: Optional[datetime] = Query(default=None),
    payment_method: Optional[str] = Query(default=None),
    include_voided: bool = Query(default=False),
    limit: int = Query(default=500, ge=1, le=2000),
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user),
):
    """List transactions with optional filters."""
    trans = TransaccionService.list_transactions(
        db,
        date_from=date_from,
        date_to=date_to,
        payment_method=payment_method,
        include_voided=include_voided,
        limit=limit,
    )
    return [_trans_to_dto(t) for t in trans]


@router.get("/reserva/{reserva_id}", summary="Transacciones + saldo de una reserva")
def get_by_reserva(
    reserva_id: str,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user),
):
    """Get all active transactions and saldo for a reservation."""
    saldo = TransaccionService.get_saldo(db, reserva_id)
    if not saldo:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Reserva no encontrada")

    saldo["transacciones"] = [_trans_to_dto(t) for t in saldo["transacciones"]]
    return saldo
